File: 429_N_tree_level_order/N_tree_level_order.cpp
/*
 * @Date: 2022-04-08 01:51:00
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2022-04-08 03:06:30
 * @FilePath: /algorithm/429_N_tree_level_order/N_tree_level_order.cpp
 */

#include <cassert>
#include <queue>
#include <vector>

using namespace std;

// Definition for a Node.
class Node {
 public:
  int val;
  vector<Node*> children;
  Node() {}
  Node(int _val) { val = _val; }
  Node(int _val, vector<Node*> _children) {
    val = _val;
    children = _children;
  }
};

class Solution {
 public:
  vector<vector<int>> levelOrder(Node* root) {
    if (!root) {
      return {};
    }

    vector<vector<int>> ans;
    queue<Node*> q;
    q.push(root);

    while (!q.empty()) {
      int cnt = q.size();
      vector<int> level;
      for (int i = 0; i < cnt; ++i) {
        Node* cur = q.front();
        q.pop();
        level.push_back(cur->val);
        for (Node* child : cur->children) {
          q.push(child);
        }
      }
      ans.push_back(move(level));
    }

    return ans;
  }
};

int main() {
  assert((
      Solution().levelOrder(new Node{
          1, vector<Node*>{new Node{3, vector<Node*>{new Node{5}, new Node{6}}},
                           new Node{2}, new Node{4}}}) ==
      vector<vector<int>>{{1}, {3, 2, 4}, {5, 6}}));

  assert((
      Solution().levelOrder((new Node{
          1,
          vector<Node*>{
              new Node{2},
              new Node{
                  3,
                  vector<Node*>{
                      new Node{6},
                      new Node{7,
                               vector<Node*>{
                                   new Node{11, vector<Node*>{new Node{14}}},
                               }}}},
              new Node{4,
                       vector<Node*>{new Node{8, vector<Node*>{new Node{12}}}}},
              new Node{5,
                       vector<Node*>{new Node{9, vector<Node*>{new Node{13}}},
                                     new Node{10}}}}})) ==
      vector<vector<int>>{
          {1}, {2, 3, 4, 5}, {6, 7, 8, 9, 10}, {11, 12, 13}, {14}}));
}

File: golang/429_N_tree_level_order/N_tree_level_order.go
/*
 * @Date: 2022-04-08 01:51:08
 * @Author: Mengsen Wang
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2024-02-17
 * @FilePath: /algorithm/golang/429_N_tree_level_order/N_tree_level_order.go
 */

package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

// Definition for a Node.
type Node struct {
	Val      int
	Children []*Node
}

func levelOrder(root *Node) (ans [][]int) {
	if root == nil {
		return
	}
	q := []*Node{root}
	for q != nil {
		level := []int{}
		tmp := q
		q = nil
		for _, node := range tmp {
			level = append(level, node.Val)
			q = append(q, node.Children...)
		}
		ans = append(ans, level)
	}
	return
}

func main() {
	tests := []struct {
		root *Node
		ans  [][]int
	}{
		{&Node{Val: 1, Children: []*Node{{Val: 3, Children: []*Node{{Val: 5, Children: nil}, {Val: 6, Children: nil}}}, {Val: 2, Children: nil}, {Val: 4, Children: nil}}}, [][]int{{1}, {3, 2, 4}, {5, 6}}},
		{&Node{Val: 1, Children: []*Node{{Val: 2, Children: nil}, {Val: 3, Children: []*Node{{Val: 6, Children: nil}, {Val: 7, Children: []*Node{{Val: 11, Children: []*Node{{Val: 14, Children: nil}}}}}}}, {Val: 4, Children: []*Node{{Val: 8, Children: []*Node{{Val: 12, Children: nil}}}}}, {Val: 5, Children: []*Node{{Val: 9, Children: []*Node{{Val: 13, Children: nil}}}, {Val: 10, Children: nil}}}}}, [][]int{{1}, {2, 3, 4, 5}, {6, 7, 8, 9, 10}, {11, 12, 13}, {14}}},
	}
	for index, test := range tests {
		assert.Equal(&testing.T{}, test.ans, levelOrder(test.root), index)
	}
}
